src/directory_listing.rs

```diff
@@ -435,30 +434,34 @@ fn html_auto_index<'a>(
     );
 
     // Prepare table row template
-    let mut table_row = String::new();
-    if base_path != "/" {
-        table_row = String::from(r#"<tr><td colspan="3"><a href="../">../</a></td></tr>"#);
-    }
+    let table_rows = if base_path == "/" {
+        String::new()
+    } else {
+        String::from(r#"<tr><td colspan="3"><a href="../">../</a></td></tr>"#)
+    };
 
-    for entry in entries {
-        let file_name = &entry.name_encoded;
+    let table_rows = entries.iter().fold(table_rows, |mut output, entry| {
+        let file_name = &entry.name;
+        let file_name_suffix = if entry.is_dir { "/" } else { "" };
         let file_modified = &entry.modified;
-        let file_uri = &entry.uri.clone().unwrap_or_else(|| file_name.to_owned());
-        let file_name_decoded = percent_decode_str(file_name).decode_utf8()?.to_string();
```

Comment:
Are you sure if the `percent_decode_str` is not necessary here, especially in environments like Windows where files/dirs can likely contain special chars? I guess we also have to check if those names are displayed correctly if not yet.

Reply:
Also, even an integration test case in `tests/dir_listing.rs` should be not bad.

Reply:
Yes, `percent_decode_str` is definitely not necessary given that we use `entry.name` rather than `entry.name_encoded` – that call was merely reverting encoding we’ve done previously. File names not using proper UTF-8 encoding will fail earlier, in `read_dir_entries()` where we try to attempt converting paths into UTF-8 strings.

In fact, the correct approach would be using `OsString` for `entry.name` here and only performing string conversion for displaying the file name (using `to_string_lossy()` then to deal with invalid encodings). Also, `percent_encode` should be called instead of `utf8_percent_encode`. But dealing with non-UTF8 file names is something for the next PR, I suspect that the code actually loading them also has issues.

Reply:
> Also, even an integration test case in `tests/dir_listing.rs` should be not bad.

Actually, there is a test already: https://github.com/static-web-server/static-web-server/blob/ddda871871ea917d92b05741e4b1bbba7c27d42f/tests/dir_listing.rs#L216

I don’t see what else I could test here, other than a file name which is non-UTF8. But while git appears to support such file names, checking in such a file might cause issues somewhere. And currently this test cannot be expected to pass given the issue I mentioned above.

Reply:
> Yes, `percent_decode_str` is definitely not necessary given that we use `entry.name` rather than `entry.name_encoded` – that call was merely reverting encoding we’ve done previously. File names not using proper UTF-8 encoding will fail earlier, in `read_dir_entries()` where we try to attempt converting paths into UTF-8 strings.

Ok, in that regard.
A user [reported](https://github.com/static-web-server/static-web-server/issues/332#issuecomment-2041646442) that we are encoding unreserved marks for the file names. For me, this is a bug so it would be perfect for another fix PR. Let me know if you want to also take care of it.

Reply:
> > Yes, `percent_decode_str` is definitely not necessary given that we use `entry.name` rather than `entry.name_encoded` – that call was merely reverting encoding we’ve done previously. File names not using proper UTF-8 encoding will fail earlier, in `read_dir_entries()` where we try to attempt converting paths into UTF-8 strings.
> 
> Ok, in that regard. A user [reported](https://github.com/static-web-server/static-web-server/issues/332#issuecomment-2041646442) that we are encoding unreserved marks for the file names. For me, this is a bug so it would be perfect for another fix PR. Let me know if you want to also take care of it.

It is not that the whole URI is encoded but the file name.
Just correcting my previous comment on https://github.com/static-web-server/static-web-server/issues/332#issuecomment-2041646442.